swarm/network/stream/delivery_test.go

```diff
@@ -302,26 +289,28 @@ func TestStreamerDownstreamChunkDeliveryMsgExchange(t *testing.T) {
 
 }
 
-func TestDeliveryFromNodes(t *testing.T) {
-	testDeliveryFromNodes(t, 2, dataChunkCount, true)
-	testDeliveryFromNodes(t, 2, dataChunkCount, false)
-	testDeliveryFromNodes(t, 4, dataChunkCount, true)
-	testDeliveryFromNodes(t, 4, dataChunkCount, false)
+// TestDeliveryFromNodes adds N nodes in a chain, meaning every node has 2 peers, and both those peers
+// are in its nearest neighbourhood. This means when a node handles a retrieve request from one peer,
+// it should always forward that request to the next peer, no matter what the chunk address is.
+// Then we randomly upload dataChunkCount chunks and try to retrieve them from one `pivot` node.
+func XTestDeliveryFromNodes(t *testing.T) {
```

Comment:
Because this test is adding nodes in a chain, and requests are sent from first peer to second peer. This is not how `FindPeer` suggests peers anymore, so this test is no longer applicable.

Reply:
My analysis (may be wrong):

* The test uploads chunks in round robin to all nodes in the simulation (one after the other)
* Nodes are connected in a chain
* Every node X is thus connected to 2 other nodes which are both in Xs NN
* The assumption was (from its GoDoc) ...`This means when a node handles a retrieve request from one peer, it should always forward that request to the next peer, no matter what the chunk address is`
* As I understand though, the pivot node will try to get all hashes for a given test file. But it is always the pivot node which is the origin of the request
* Therefore, in the new fetcher design, some hashes will be found and others not, because the pivot node will issue the request to its two neighbors, which, based on the chunk's address, will NOT forward every request based on the new rules for forwarding


For me this has the following implications and scenarios:


* The test is badly designed and irrelevant, as retrieving in a chain of nodes is made up and not realistic. If that is the case, the test should be removed
* The new fetcher design has the implication that retrieval will only work in a healthy network. This may be in fact desired, but we should be aware of this limitation
* My intuition suggests to me that, if I understand it correctly (and I am absolutely not sure about it), the new fetcher design is too restrictive. It should just forward the request along kademlia paths, just like any kademlia network. ignoring rules related to NN. I suggest tackling redundant requests is a different issue and should not be handled by analyzing the chunk address, but by adding a request ID, and every open fetcher is stored with this request ID, so loops should be easily be avoided like this

Reply:
Your analysis is absolutely correct and on point:
1. The test was designed to check if the forwarding of requests (when you have 2 peers), from one peer to the other, is working.
2. We needed `hopcount` to be set to a very high value for this test to pass - 16 or 17 for the 16 node network. `hopcount` has now been removed.
3. Yes, retrieval will work only in a healthy network. If chunks are not where they are supposed to be, retrieval will fail.
4. I wouldn't say that the algorithm with selects `next peer` (or as previously called `eligible peer`) is too restrictive - we follow the Kademlia addressing until we reach the NN. At that point, if a node receives a request for a chunk from other node that is closer to the chunk, then it will handle the request, but not forward it further - essentially we get rid of circular requests and the need for `hopcount`.